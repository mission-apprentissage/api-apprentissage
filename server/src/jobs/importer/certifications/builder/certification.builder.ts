import { ICertification, zCertification } from "api-alternance-sdk";
import { IBcn_N_FormationDiplome } from "shared/models/source/bcn/bcn.n_formation_diplome.model";
import { IBcn_N51_FormationDiplome } from "shared/models/source/bcn/bcn.n51_formation_diplome.model";
import { ISourceFranceCompetence } from "shared/models/source/france_competence/source.france_competence.model";

import { buildCertificationBaseLegale } from "./base_legale/certification.base_legale.builder.js";
import { buildCertificationBlocsCompetences } from "./blocs_competences/certification.blocs_competences.builder.js";
import { buildCertificationConventionCollectives } from "./convention_collectives/certification.convention_collectives.builder.js";
import { buildCertificationDomaines } from "./domaines/certification.domaines.builder.js";
import { buildCertificationIntitule } from "./intitule/certification.intitule.builder.js";
import { buildCertificationPeriodeValidite } from "./periode_validite/certification.periode_validite.builder.js";
import { buildCertificationType } from "./type/certification.type.builder.js";

export type ISourceAggregatedData = {
  bcn?: IBcn_N51_FormationDiplome | IBcn_N_FormationDiplome | null;
  france_competence?: ISourceFranceCompetence | null;
};

function isRncpAnterieur2019(rncp: string | null): boolean | null {
  if (rncp === null) {
    return null;
  }

  const numero = parseInt(rncp.substring(4));
  return numero < 34000;
}

export function buildCertification(
  data: ISourceAggregatedData,
  oldestFranceCompetenceDatePublication: Date
): ICertification {
  const cfdCode = data.bcn?.data.FORMATION_DIPLOME || null;
  const rncpCode = data.france_competence?.numero_fiche || null;

  return zCertification.parse({
    identifiant: {
      cfd: cfdCode,
      rncp: rncpCode,
      rncp_anterieur_2019: isRncpAnterieur2019(rncpCode),
    },
    intitule: buildCertificationIntitule(data),
    base_legale: buildCertificationBaseLegale(data),
    blocs_competences: buildCertificationBlocsCompetences(data),
    convention_collectives: buildCertificationConventionCollectives(data),
    domaines: buildCertificationDomaines(data),
    periode_validite: buildCertificationPeriodeValidite(data, oldestFranceCompetenceDatePublication),
    type: buildCertificationType(data),
    continuite: { rncp: null, cfd: null },
  });
}
