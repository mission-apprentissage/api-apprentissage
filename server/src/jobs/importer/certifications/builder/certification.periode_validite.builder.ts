import { ICertification } from "shared/models/certification.model";

export function buildCertificationPeriodeValidite(
  cfd: ICertification["cfd"],
  rncp: ICertification["rncp"]
): ICertification["periode_validite"] {
  const ouverture = cfd?.ouverture ?? cfd?.creation ?? null;
  const activation = rncp?.activation ?? null;

  let debut: Date | null = ouverture ?? activation;

  if (ouverture != null && activation != null) {
    debut = ouverture > activation ? ouverture : activation;
  }

  const fermeture = cfd?.fermeture ?? null;
  const finEnregistrement = rncp?.fin_enregistrement ?? null;

  let fin: Date | null = fermeture ?? finEnregistrement;
  if (fermeture != null && finEnregistrement != null) {
    fin = fermeture < finEnregistrement ? fermeture : finEnregistrement;
  }

  return {
    debut,
    fin,
  };
}
