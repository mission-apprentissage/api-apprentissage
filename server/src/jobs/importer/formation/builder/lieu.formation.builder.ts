import { internal } from "@hapi/boom";
import type { ICommune, IFormation } from "api-alternance-sdk";
import { zSiret } from "api-alternance-sdk";
import { LRUCache } from "lru-cache";
import type { IFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

import { getDbCollection } from "@/services/mongodb/mongodbService.js";

const cache = new LRUCache<string, ICommune>({
  max: 5_000,
  ttl: 60 * 60_000,
});

async function getCommune(code_insee: string): Promise<ICommune | null> {
  if (cache.has(code_insee)) {
    return cache.get(code_insee)!;
  }

  const commune = await getDbCollection("commune").findOne({
    $or: [{ "code.insee": code_insee }, { "arrondissements.code": code_insee }, { "anciennes.codeInsee": code_insee }],
  });

  if (commune) {
    cache.set(code_insee, commune);
    return commune;
  }

  const byCp = await getDbCollection("commune").findOne({ "code.postaux": code_insee });

  if (byCp) {
    cache.set(code_insee, byCp);
    return byCp;
  }

  return null;
}

function parseGeoCoordinates(geo: string | null): { lat: number; long: number } | null {
  if (!geo) {
    return null;
  }

  const [lat, long] = geo.split(",").map(Number);

  if (isNaN(lat) || isNaN(long)) {
    return null;
  }

  return { lat, long };
}

export async function buildFormationLieu(
  data: Pick<
    IFormationCatalogue,
    | "code_commune_insee"
    | "lieu_formation_geo_coordonnees"
    | "lieu_formation_geo_coordonnees_computed"
    | "lieu_formation_adresse"
    | "code_postal"
    | "distance"
    | "lieu_formation_siret"
  >
): Promise<IFormation["lieu"]> {
  let commune = await getCommune(data.code_commune_insee);

  if (!commune) {
    commune = await getCommune(data.code_postal);

    if (!commune) {
      throw internal(`buildFormationLieu: commune not found`, {
        code_commune_insee: data.code_commune_insee,
        code_postal: data.code_postal,
      });
    }
  }

  const coords =
    parseGeoCoordinates(data.lieu_formation_geo_coordonnees) ??
    parseGeoCoordinates(data.lieu_formation_geo_coordonnees_computed);

  if (coords === null) {
    throw internal(`buildFormationLieu: invalid geo coordinates`, {
      lieu_formation_geo_coordonnees: data.lieu_formation_geo_coordonnees,
      lieu_formation_geo_coordonnees_computed: data.lieu_formation_geo_coordonnees_computed,
    });
  }

  const siretParse = zSiret.safeParse(data.lieu_formation_siret);

  return {
    adresse: {
      label: data.lieu_formation_adresse,
      code_postal: data.code_postal,

      commune: {
        nom: commune.nom,
        code_insee: commune.code.insee,
      },
      departement: {
        nom: commune.departement.nom,
        code_insee: commune.departement.codeInsee,
      },
      region: {
        code_insee: commune.region.codeInsee,
        nom: commune.region.nom,
      },
      academie: {
        id: commune.academie.id,
        code: commune.academie.code,
        nom: commune.academie.nom,
      },
    },

    geolocalisation: {
      type: "Point",
      coordinates: [coords.long, coords.lat],
    },

    precision: data.distance,

    siret: siretParse.success ? siretParse.data : null,
  };
}
