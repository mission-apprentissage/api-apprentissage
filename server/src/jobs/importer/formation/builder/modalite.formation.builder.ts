import { internal } from "@hapi/boom";
import type { IFormation } from "api-alternance-sdk";
import type { IFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";
import { z } from "zod/v4-mini";

export function buildFormationModalite(
  data: Pick<IFormationCatalogue, "duree" | "entierement_a_distance" | "annee" | "bcn_mefs_10">
): IFormation["modalite"] {
  const duree = z.safeParse(z.coerce.number().check(z.positive(), z.int()), data.duree);

  if (!duree.success) {
    throw internal(`buildModalite: invalid duree`, { duree: data.duree });
  }

  if (data.annee === "X") {
    return {
      entierement_a_distance: data.entierement_a_distance,
      duree_indicative: duree.data,
      annee_cycle: null,
      mef_10: data.bcn_mefs_10.length === 1 ? data.bcn_mefs_10[0].mef10 : null,
    };
  }

  const annee = z.safeParse(z.coerce.number().check(z.positive(), z.int()), data.annee);
  if (!annee.success) {
    throw internal(`buildModalite: invalid annee`, { annee: data.annee });
  }

  return {
    entierement_a_distance: data.entierement_a_distance,
    duree_indicative: duree.data,
    annee_cycle: annee.data,
    mef_10: data.bcn_mefs_10.length === 1 ? data.bcn_mefs_10[0].mef10 : null,
  };
}
