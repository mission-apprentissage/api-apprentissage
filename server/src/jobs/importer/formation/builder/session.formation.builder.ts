import { internal } from "@hapi/boom";
import type { IFormation } from "api-alternance-sdk";
import { ParisDate } from "api-alternance-sdk/internal";
import { Interval } from "luxon";
import type { IFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

type IModaliteData = Pick<IFormation["modalite"], "annee_cycle" | "duree_indicative">;

function getBestMatchSession(sessions: { debut: ParisDate; fin: ParisDate }[], modalite: IModaliteData) {
  let bestMatch = null;
  let bestMatchScore = 0;

  const annee_cycle = modalite.annee_cycle ?? 1;
  const expectedDuration = modalite.duree_indicative - annee_cycle + 1;

  for (const session of sessions) {
    const interval = Interval.fromDateTimes(session.debut, session.fin);
    if (!interval.isValid) {
      continue;
    }
    const duration = interval.toDuration(["years"]).years;
    const score = Math.abs(expectedDuration - duration);

    if (bestMatch === null || score < bestMatchScore) {
      bestMatch = session;
      bestMatchScore = score;
    }
  }

  if (bestMatch === null) {
    throw internal(`getBestMatchSession: no best match found`);
  }

  return bestMatch;
}

export function buildFormationSessions(
  data: Pick<IFormationCatalogue, "date_debut" | "date_fin" | "capacite">,
  modalite: IModaliteData
): IFormation["sessions"] {
  const dateDebuts = data.date_debut?.toSorted() ?? [];
  const dateFins = data.date_fin?.toSorted() ?? [];

  const sessions: IFormation["sessions"] = [];

  if (dateDebuts.length === 0 || dateFins.length === 0) {
    return sessions;
  }

  if (dateDebuts.length === dateFins.length) {
    for (let i = 0; i < dateDebuts.length; i++) {
      sessions.push({
        debut: new ParisDate(dateDebuts[i]),
        fin: new ParisDate(dateFins[i]),
        capacite: data.capacite == null ? null : Number(data.capacite),
      });
    }
  }

  if (dateDebuts.length > dateFins.length) {
    for (let i = 0; i < dateFins.length; i++) {
      const fin = new ParisDate(dateFins[i]);

      const bestMatch = getBestMatchSession(
        dateDebuts.map((dateDebut) => ({
          debut: new ParisDate(dateDebut),
          fin,
        })),
        modalite
      );

      sessions.push({
        ...bestMatch,
        capacite: data.capacite == null ? null : Number(data.capacite),
      });
    }
  }

  if (dateDebuts.length < dateFins.length) {
    for (let i = 0; i < dateDebuts.length; i++) {
      const debut = new ParisDate(dateDebuts[i]);

      const bestMatch = getBestMatchSession(
        dateFins.map((dateFin) => ({
          debut,
          fin: new ParisDate(dateFin),
        })),
        modalite
      );

      sessions.push({
        ...bestMatch,
        capacite: data.capacite == null ? null : Number(data.capacite),
      });
    }
  }

  return sessions;
}
