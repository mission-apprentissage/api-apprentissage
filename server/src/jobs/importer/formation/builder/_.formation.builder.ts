import type { IFormation } from "api-alternance-sdk";
import { zFormation } from "api-alternance-sdk";
import type { ISourceCatalogue } from "shared/models/source/catalogue/source.catalogue.model";
import { z } from "zod/v4-mini";

import { buildFormationCertification } from "./certification.formation.builder.js";
import { buildFormationLieu } from "./lieu.formation.builder.js";
import { buildFormationModalite } from "./modalite.formation.builder.js";
import { buildFormationOrganisme } from "./organisme.formation.builder.js";
import { buildFormationSessions } from "./session.formation.builder.js";

export async function buildFormation(source: ISourceCatalogue): Promise<IFormation> {
  const [lieu, certification, formateur, responsable] = await Promise.all([
    buildFormationLieu(source.data),
    buildFormationCertification(source.data),
    buildFormationOrganisme({
      siret: source.data.etablissement_formateur_siret,
      uai: source.data.etablissement_formateur_uai,
    }),
    buildFormationOrganisme({
      siret: source.data.etablissement_gestionnaire_siret,
      uai: source.data.etablissement_gestionnaire_uai,
    }),
  ]);

  const modalite = buildFormationModalite(source.data);
  const sessions = buildFormationSessions(source.data, modalite);

  const data: IFormation = {
    identifiant: {
      cle_ministere_educatif: source.data.cle_ministere_educatif,
    },

    statut: {
      catalogue: source.data.published ? "publié" : "archivé",
    },

    formateur,

    responsable,

    certification,

    lieu,

    contact: {
      // Some emails are not valid
      email: z.string().check(z.email()).safeParse(source.data.email).success ? source.data.email : null,
      telephone: source.data.num_tel,
    },

    onisep: {
      // Some onisep are not valid URL
      url: z.string().check(z.url()).safeParse(source.data.onisep_url).success ? source.data.onisep_url : null,
      intitule: source.data.onisep_intitule,
      libelle_poursuite: source.data.onisep_libelle_poursuite,
      lien_site_onisepfr: source.data.onisep_lien_site_onisepfr,
      discipline: source.data.onisep_discipline,
      domaine_sousdomaine: source.data.onisep_domaine_sousdomaine,
    },

    modalite,

    contenu_educatif: {
      contenu: source.data.contenu ?? "",
      objectif: source.data.objectif ?? "",
    },

    sessions,
  };

  return zFormation.parse(data);
}
