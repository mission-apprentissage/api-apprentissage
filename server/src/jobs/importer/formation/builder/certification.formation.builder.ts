import { internal } from "@hapi/boom";
import type { ICertification, IFormation } from "api-alternance-sdk";
import { LRUCache } from "lru-cache";
import { stringify } from "safe-stable-stringify";
import type { IFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

import { computePeriodeValidite } from "@/jobs/importer/certifications/builder/periode_validite/certification.periode_validite.builder.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";

const cache = new LRUCache<string, IFormation["certification"]>({
  max: 5_000,
  ttl: 60 * 60_000,
});

async function getCertificationFromRncp(code: string | null): Promise<ICertification | null> {
  if (!code || code === "RNCPNR") {
    return null;
  }

  const certification = await getDbCollection("certifications").findOne({ "identifiant.rncp": code });

  if (!certification) {
    throw internal(`getCertificationFromRncp: certification not found for code`, { code });
  }

  return certification;
}

async function getCertificationFromCfd(code: string): Promise<ICertification> {
  const certification = await getDbCollection("certifications").findOne({ "identifiant.cfd": code });

  if (!certification) {
    throw internal(`getCertificationFromCfd: certification not found for code`, { code });
  }

  return certification;
}

export async function buildFormationCertification(
  data: Pick<IFormationCatalogue, "rncp_code" | "cfd">
): Promise<IFormation["certification"]> {
  const identifiant = {
    rncp: data.rncp_code,
    cfd: data.cfd,
  };

  const cacheKey = stringify(identifiant);

  if (cache.has(cacheKey)) {
    return cache.get(cacheKey)!;
  }

  const certification = await getDbCollection("certifications").findOne({
    "identifiant.rncp": identifiant.rncp,
    "identifiant.cfd": identifiant.cfd,
  });

  if (certification) {
    const result = {
      valeur: certification,
      connue: true,
    };

    cache.set(cacheKey, result);
    return result;
  }

  const [rncpCertif, cfdCertif] = await Promise.all([
    getCertificationFromRncp(data.rncp_code),
    getCertificationFromCfd(data.cfd),
  ]);

  const result = {
    valeur: {
      identifiant: {
        rncp: rncpCertif?.identifiant.rncp ?? null,
        cfd: cfdCertif.identifiant.cfd,
        rncp_anterieur_2019: rncpCertif?.identifiant.rncp_anterieur_2019 ?? null,
      },
      intitule: {
        cfd: cfdCertif.intitule.cfd,
        rncp: rncpCertif?.intitule.rncp ?? null,
        niveau: {
          cfd: cfdCertif.intitule.niveau.cfd,
          rncp: rncpCertif?.intitule.niveau.rncp ?? null,
        },
      },
      base_legale: {
        cfd: cfdCertif.base_legale.cfd,
      },
      blocs_competences: {
        rncp: rncpCertif?.blocs_competences.rncp ?? null,
      },
      convention_collectives: {
        rncp: rncpCertif?.convention_collectives.rncp ?? null,
      },
      domaines: {
        formacodes: {
          rncp: rncpCertif?.domaines.formacodes.rncp ?? null,
        },
        nsf: {
          cfd: cfdCertif.domaines.nsf.cfd,
          rncp: rncpCertif?.domaines.nsf.rncp ?? null,
        },
        rome: {
          rncp: rncpCertif?.domaines.rome.rncp ?? null,
        },
      },
      periode_validite: computePeriodeValidite(
        cfdCertif.periode_validite.cfd,
        rncpCertif?.periode_validite.rncp ?? null
      ),
      type: {
        nature: {
          cfd: cfdCertif.type.nature.cfd,
        },
        gestionnaire_diplome: cfdCertif.type.gestionnaire_diplome,
        enregistrement_rncp: rncpCertif?.type.enregistrement_rncp ?? null,
        voie_acces: {
          rncp: rncpCertif?.type.voie_acces.rncp ?? null,
        },
        certificateurs_rncp: rncpCertif?.type.certificateurs_rncp ?? null,
      },
      continuite: {
        cfd: cfdCertif.continuite.cfd,
        rncp: rncpCertif?.continuite.rncp ?? null,
      },
    },
    connue: false,
  };

  cache.set(cacheKey, result);
  return result;
}
