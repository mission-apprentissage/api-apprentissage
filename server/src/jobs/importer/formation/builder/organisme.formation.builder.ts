import type { IFormation, IOrganisme } from "api-alternance-sdk";
import { zSiret, zUai } from "api-alternance-sdk";
import { LRUCache } from "lru-cache";
import { stringify } from "safe-stable-stringify";

import { buildOrganisme, buildOrganismeContext } from "@/jobs/importer/organisme/builder/organisme.builder.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";

const cache = new LRUCache<string, Promise<IFormation["formateur"]>>({
  max: 5_000,
  ttl: 60 * 60_000,
});

async function buildFormationOrganismeRaw(identifiant: IOrganisme["identifiant"]): Promise<IFormation["formateur"]> {
  const organisme = await getDbCollection("organisme").findOne({
    "identifiant.siret": identifiant.siret,
    "identifiant.uai": identifiant.uai?.toUpperCase() ?? null,
  });

  if (organisme) {
    return {
      organisme: organisme,
      connu: true,
    };
  }

  const bySiret = await getDbCollection("organisme").findOne({
    "identifiant.siret": identifiant.siret,
  });

  if (bySiret) {
    return {
      organisme: bySiret,
      connu: true,
    };
  }

  if (!zSiret.safeParse(identifiant.siret).success) {
    return {
      organisme: null,
      connu: false,
    };
  }
  const context = await buildOrganismeContext(identifiant.siret);

  if (!context) {
    return {
      organisme: null,
      connu: false,
    };
  }

  return {
    organisme: buildOrganisme(
      {
        siret: identifiant.siret,
        uai: zUai.safeParse(identifiant.uai).success ? identifiant.uai : null,
      },
      context,
      "supprim√©"
    ),
    connu: false,
  };
}

export async function buildFormationOrganisme(
  identifiant: IOrganisme["identifiant"]
): Promise<IFormation["formateur"]> {
  const cacheKey = stringify(identifiant);
  if (cache.has(cacheKey)) {
    return cache.get(cacheKey)!;
  }

  const result = buildFormationOrganismeRaw(identifiant);
  cache.set(cacheKey, result);
  return result;
}
