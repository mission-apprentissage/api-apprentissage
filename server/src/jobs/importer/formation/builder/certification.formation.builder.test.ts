import { useMongo } from "@tests/mongo.test.utils.js";
import { generateCertificationFixture } from "shared/models/fixtures/certification.model.fixture";
import { beforeEach, describe, expect, it } from "vitest";

import { getDbCollection } from "@/services/mongodb/mongodbService.js";

import { buildFormationCertification } from "./certification.formation.builder.js";

useMongo();

describe("buildFormationCertification", () => {
  const certifications = [
    generateCertificationFixture({
      identifiant: {
        cfd: "46125123",
        rncp: null,
        rncp_anterieur_2019: null,
      },
      intitule: {
        cfd: {
          long: "MAINTENANCE DES INSTALLATIONS OLEOHYDRAULIQUES ET PNEUMATIQUES (CS4)",
          court: "MAINTEN.INSTAL.OLEOHYDRAU.PNEUM.",
        },
        niveau: {
          cfd: {
            europeen: "4",
            formation_diplome: "461",
            interministeriel: "4",
            libelle: "DIPLOME NATIONAL / DIPLOME D'ETAT",
            sigle: "CS4",
          },
          rncp: null,
        },
        rncp: null,
      },
      base_legale: {
        cfd: {
          creation: new Date("2023-08-24T22:00:00.000Z"),
          abrogation: null,
        },
      },
      blocs_competences: {
        rncp: null,
      },
      convention_collectives: {
        rncp: null,
      },
      domaines: {
        formacodes: {
          rncp: null,
        },
        nsf: {
          cfd: {
            code: "251",
            intitule: "MECANIQ.GENERALE & DE PRECISION, USINAGE",
          },
          rncp: null,
        },
        rome: {
          rncp: null,
        },
      },
      periode_validite: {
        debut: new Date("2026-08-31T22:00:00.000Z"),
        fin: null,
        cfd: {
          ouverture: new Date("2024-08-31T22:00:00.000Z"),
          fermeture: null,
          premiere_session: 2025,
          derniere_session: null,
        },
        rncp: null,
      },
      type: {
        nature: {
          cfd: {
            code: "1",
            libelle: "DIPLOME NATIONAL / DIPLOME D'ETAT",
          },
        },
        gestionnaire_diplome: "DGESCO A2-3",
        enregistrement_rncp: null,
        voie_acces: {
          rncp: null,
        },
        certificateurs_rncp: null,
      },
      continuite: {
        cfd: [
          {
            code: "01025116",
            ouverture: new Date("1999-08-31T22:00:00.000Z"),
            fermeture: new Date("2003-08-31T21:59:59.000Z"),
            courant: false,
          },
          {
            code: "01025123",
            ouverture: new Date("2002-08-31T22:00:00.000Z"),
            fermeture: new Date("2024-08-31T21:59:59.000Z"),
            courant: false,
          },
          {
            code: "46125123",
            ouverture: new Date("2024-08-31T22:00:00.000Z"),
            fermeture: null,
            courant: true,
          },
        ],
        rncp: null,
      },
      created_at: new Date("2025-01-02T12:09:58.075Z"),
      updated_at: new Date("2025-01-02T12:09:58.075Z"),
    }),
    generateCertificationFixture({
      identifiant: {
        rncp: "RNCP35234",
        cfd: "56T34302",
        rncp_anterieur_2019: false,
      },
      base_legale: {
        cfd: {
          creation: null,
          abrogation: new Date("2027-02-08T22:59:59.000Z"),
        },
      },
      blocs_competences: {
        rncp: [
          {
            code: "RNCP35234BC02",
            intitule: "Réaliser des opérations d'exploitation sur un site d'apport volontaire de déchets",
          },
          {
            code: "RNCP35234BC01",
            intitule: "Accueillir les usagers/clients sur un site d'apport volontaire de déchets",
          },
        ],
      },
      continuite: {
        cfd: null,
        rncp: [
          {
            code: "RNCP34501",
            activation: null,
            fin_enregistrement: new Date("2022-02-08T22:59:59.000Z"),
            courant: false,
            actif: false,
          },
          {
            code: "RNCP4794",
            activation: null,
            fin_enregistrement: new Date("2020-02-08T22:59:59.000Z"),
            courant: false,
            actif: false,
          },
          {
            code: "RNCP35234",
            activation: null,
            fin_enregistrement: new Date("2027-02-08T22:59:59.000Z"),
            courant: true,
            actif: true,
          },
        ],
      },
      convention_collectives: {
        rncp: [],
      },
      created_at: new Date("2024-12-30T10:21:44.050Z"),
      domaines: {
        formacodes: {
          rncp: [
            {
              code: "12582",
              intitule: "12582 : Gestion déchet",
            },
          ],
        },
        nsf: {
          cfd: {
            code: "343",
            intitule: "NETTOYAGE, ASSAIN., PROTECTION ENVIRONMT",
          },
          rncp: [
            {
              code: "343",
              intitule: "343 : Nettoyage, assainissement, protection de l'environnement",
            },
          ],
        },
        rome: {
          rncp: [
            {
              code: "K2304",
              intitule: "Revalorisation de produits industriels",
            },
            {
              code: "K2303",
              intitule: "Nettoyage des espaces urbains",
            },
          ],
        },
      },
      intitule: {
        cfd: {
          long: "AGENT TECHNIQUE DE RECEPTION ET DE VALORISATION DES DECHETS (TP)",
          court: "AGT TECHN RECEPT VALORISA DECHETS",
        },
        niveau: {
          cfd: {
            europeen: "3",
            formation_diplome: "56T",
            interministeriel: "5",
            libelle: "TITRE PROFESSIONNEL HOMOLOGUE OU CERTIFIE",
            sigle: "TH5-T",
          },
          rncp: {
            europeen: "3",
          },
        },
        rncp: "Agent technique de réception et de valorisation de déchets",
      },
      periode_validite: {
        debut: new Date("2017-07-31T22:00:00.000Z"),
        fin: new Date("2027-02-08T22:59:59.000Z"),
        cfd: {
          ouverture: new Date("2017-07-31T22:00:00.000Z"),
          fermeture: new Date("2027-08-31T21:59:59.000Z"),
          premiere_session: null,
          derniere_session: null,
        },
        rncp: {
          actif: true,
          activation: null,
          debut_parcours: new Date("2022-02-07T23:00:00.000Z"),
          fin_enregistrement: new Date("2027-02-08T22:59:59.000Z"),
        },
      },
      type: {
        nature: {
          cfd: {
            code: "2",
            libelle: "TITRE PROFESSIONNEL HOMOLOGUE OU CERTIFIE",
          },
        },
        gestionnaire_diplome: "DEPPA1",
        enregistrement_rncp: "Enregistrement de droit",
        voie_acces: {
          rncp: {
            apprentissage: true,
            experience: true,
            candidature_individuelle: false,
            contrat_professionnalisation: true,
            formation_continue: true,
            formation_statut_eleve: false,
          },
        },
        certificateurs_rncp: [
          {
            siret: "11000007200014",
            nom: "MINISTERE DU TRAVAIL DU PLEIN EMPLOI ET DE L' INSERTION",
          },
        ],
      },
      updated_at: new Date("2025-01-03T05:00:06.642Z"),
    }),
  ];

  beforeEach(async () => {
    await getDbCollection("certifications").insertMany(certifications);
  });

  it("should return certification when it exists in db", async () => {
    const result = await buildFormationCertification({
      cfd: certifications[1].identifiant.cfd ?? "",
      rncp_code: certifications[1].identifiant.rncp,
    });

    expect(result).toEqual({
      valeur: certifications[1],
      connue: true,
    });

    // Should be cached
    const cachedResult = await buildFormationCertification({
      cfd: certifications[1].identifiant.cfd ?? "",
      rncp_code: certifications[1].identifiant.rncp,
    });
    expect(cachedResult).toBe(result);
  });

  it("should build certification if the association is not known", async () => {
    const result = await buildFormationCertification({
      cfd: certifications[0].identifiant.cfd ?? "",
      rncp_code: certifications[1].identifiant.rncp,
    });

    expect(result.connue).toBe(false);
    expect(result.valeur).toMatchSnapshot();
  });

  it('should error if "cfd" is not found', async () => {
    await expect(
      buildFormationCertification({ cfd: "unknown", rncp_code: certifications[1].identifiant.rncp })
    ).rejects.toThrowError("getCertificationFromCfd: certification not found for code");
  });

  it('should error if "rncp" is not found', async () => {
    await expect(
      buildFormationCertification({ cfd: certifications[0].identifiant.cfd ?? "", rncp_code: "unknown" })
    ).rejects.toThrowError("getCertificationFromRncp: certification not found for code");
  });
});
