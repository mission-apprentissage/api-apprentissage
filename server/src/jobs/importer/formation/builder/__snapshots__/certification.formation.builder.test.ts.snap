// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildFormationCertification > should build certification if the association is not known 1`] = `
{
  "base_legale": {
    "cfd": {
      "abrogation": null,
      "creation": 2023-08-24T22:00:00.000Z,
    },
  },
  "blocs_competences": {
    "rncp": [
      {
        "code": "RNCP35234BC02",
        "intitule": "Réaliser des opérations d'exploitation sur un site d'apport volontaire de déchets",
      },
      {
        "code": "RNCP35234BC01",
        "intitule": "Accueillir les usagers/clients sur un site d'apport volontaire de déchets",
      },
    ],
  },
  "continuite": {
    "cfd": [
      {
        "code": "01025116",
        "courant": false,
        "fermeture": 2003-08-31T21:59:59.000Z,
        "ouverture": 1999-08-31T22:00:00.000Z,
      },
      {
        "code": "01025123",
        "courant": false,
        "fermeture": 2024-08-31T21:59:59.000Z,
        "ouverture": 2002-08-31T22:00:00.000Z,
      },
      {
        "code": "46125123",
        "courant": true,
        "fermeture": null,
        "ouverture": 2024-08-31T22:00:00.000Z,
      },
    ],
    "rncp": [
      {
        "actif": false,
        "activation": null,
        "code": "RNCP34501",
        "courant": false,
        "fin_enregistrement": 2022-02-08T22:59:59.000Z,
      },
      {
        "actif": false,
        "activation": null,
        "code": "RNCP4794",
        "courant": false,
        "fin_enregistrement": 2020-02-08T22:59:59.000Z,
      },
      {
        "actif": true,
        "activation": null,
        "code": "RNCP35234",
        "courant": true,
        "fin_enregistrement": 2027-02-08T22:59:59.000Z,
      },
    ],
  },
  "convention_collectives": {
    "rncp": [],
  },
  "domaines": {
    "formacodes": {
      "rncp": [
        {
          "code": "12582",
          "intitule": "12582 : Gestion déchet",
        },
      ],
    },
    "nsf": {
      "cfd": {
        "code": "251",
        "intitule": "MECANIQ.GENERALE & DE PRECISION, USINAGE",
      },
      "rncp": [
        {
          "code": "343",
          "intitule": "343 : Nettoyage, assainissement, protection de l'environnement",
        },
      ],
    },
    "rome": {
      "rncp": [
        {
          "code": "K2304",
          "intitule": "Revalorisation de produits industriels",
        },
        {
          "code": "K2303",
          "intitule": "Nettoyage des espaces urbains",
        },
      ],
    },
  },
  "identifiant": {
    "cfd": "46125123",
    "rncp": "RNCP35234",
    "rncp_anterieur_2019": false,
  },
  "intitule": {
    "cfd": {
      "court": "MAINTEN.INSTAL.OLEOHYDRAU.PNEUM.",
      "long": "MAINTENANCE DES INSTALLATIONS OLEOHYDRAULIQUES ET PNEUMATIQUES (CS4)",
    },
    "niveau": {
      "cfd": {
        "europeen": "4",
        "formation_diplome": "461",
        "interministeriel": "4",
        "libelle": "DIPLOME NATIONAL / DIPLOME D'ETAT",
        "sigle": "CS4",
      },
      "rncp": {
        "europeen": "3",
      },
    },
    "rncp": "Agent technique de réception et de valorisation de déchets",
  },
  "periode_validite": {
    "cfd": {
      "derniere_session": null,
      "fermeture": null,
      "ouverture": 2024-08-31T22:00:00.000Z,
      "premiere_session": 2025,
    },
    "debut": 2024-08-31T22:00:00.000Z,
    "fin": 2027-02-08T22:59:59.000Z,
    "rncp": {
      "actif": true,
      "activation": null,
      "debut_parcours": 2022-02-07T23:00:00.000Z,
      "fin_enregistrement": 2027-02-08T22:59:59.000Z,
    },
  },
  "type": {
    "certificateurs_rncp": [
      {
        "nom": "MINISTERE DU TRAVAIL DU PLEIN EMPLOI ET DE L' INSERTION",
        "siret": "11000007200014",
      },
    ],
    "enregistrement_rncp": "Enregistrement de droit",
    "gestionnaire_diplome": "DGESCO A2-3",
    "nature": {
      "cfd": {
        "code": "1",
        "libelle": "DIPLOME NATIONAL / DIPLOME D'ETAT",
      },
    },
    "voie_acces": {
      "rncp": {
        "apprentissage": true,
        "candidature_individuelle": false,
        "contrat_professionnalisation": true,
        "experience": true,
        "formation_continue": true,
        "formation_statut_eleve": false,
      },
    },
  },
}
`;
